mistralrs-core/src/engine/mod.rs

```diff
@@ -213,21 +243,113 @@ impl Engine {
+        logits: Tensor,
+        seq: &mut Sequence,
+        return_logprobs: bool,
+        repeat_last_n: usize,
+        tok_trie: Arc<TokTrie>,
+        rng: Arc<Mutex<Isaac64Rng>>,
+        use_async_pool: bool,
+    ) -> Result<sampler::Logprobs> {
+        let logits = logits.squeeze(0)?.squeeze(0)?.to_dtype(DType::F32)?;
+        let start_at = seq.get_toks().len().saturating_sub(repeat_last_n);
+
+        let sampler = seq.sampler();
+        let logits_clone = logits.clone();
+        let ctx_clone = seq.get_toks()[start_at..].to_vec();
+        let rng_clone = rng.clone();
+        let first_lobprobs_response = sample_async!(
+            use_async_pool,
+            sampler,
+            logits_clone,
+            ctx_clone,
+            return_logprobs,
+            rng_clone
+        );
+
+        let bias_if_not_allowed = match &mut seq.recognizer {
+            SequenceRecognizer::Regex(ref mut rx) => {
+                get_bias_if_not_allowed!(tok_trie, rx.as_mut(), first_lobprobs_response.token)
+            }
+            SequenceRecognizer::Cfg(ref mut cfg) => {
+                get_bias_if_not_allowed!(tok_trie, cfg.as_mut(), first_lobprobs_response.token)
+            }
+            SequenceRecognizer::None => None,
+        };
+        let second_logprobs_response = match bias_if_not_allowed {
+            Some(token_set) => {
+                let mut acc = vec![-f32::INFINITY; tok_trie.vocab_size()];
+                token_set.apply_to(&mut acc);
+                let new_logits = (logits + Tensor::from_slice(&acc, acc.len(), &Device::Cpu)?)?;
+
+                let ctx_clone = seq.get_toks()[start_at..].to_vec();
+                let rng_clone = rng.clone();
+                let sampler = seq.sampler();
+                sample_async!(
+                    use_async_pool,
+                    sampler,
+                    new_logits,
+                    ctx_clone,
+                    return_logprobs,
+                    rng_clone
+                )
+            }
+            None => first_lobprobs_response,
+        };
+
+        match seq.recognizer {
+            SequenceRecognizer::Regex(ref mut rx) => {
+                tok_trie.append_token(rx.as_mut(), second_logprobs_response.token);
+            }
+            SequenceRecognizer::Cfg(ref mut cfg) => {
+                tok_trie.append_token(cfg.as_mut(), second_logprobs_response.token);
+            }
+            SequenceRecognizer::None => {}
+        }
+        Ok(second_logprobs_response)
+    }
+    async fn sample_seqs(
+        pipeline: Arc<Mutex<dyn Pipeline>>,
         seqs: &mut [&mut Sequence],
         logits: Tensor,
         prefix_cacher: &mut PrefixCacheManager,
         disable_eos_stop: bool,
+        rng: Arc<Mutex<Isaac64Rng>>,
     ) -> Result<()> {
         let seqs_len = seqs.len();
-        let logits_seq = logits.chunk(seqs_len, 0).unwrap();
+        let logits_seq = logits.to_device(&Device::Cpu)?.chunk(seqs_len, 0)?;
```

Comment:
Now that we do a synchronization before we start sampling, we can get statistics about sampling speed.

Basically reverting https://github.com/EricLBuehler/mistral.rs/pull/151